auth/src/main/java/com/google/cloud/auth/samples/DownscopingExample.java

```diff
@@ -0,0 +1,150 @@
+ */
+
+package com.google.cloud.auth.samples;
+
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.CredentialAccessBoundary;
+import com.google.auth.oauth2.DownscopedCredentials;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.OAuth2CredentialsWithRefresh;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import java.io.IOException;
+
+/** Demonstrates how to use Downscoping with Credential Access Boundaries.". */
+public class DownscopingExample {
+
+  /**
+   * Tests the downscoping functionality.
+   *
+   * <p>This will generate a downscoped token with readonly access to the specified GCS bucket,
+   * inject them into a storage instance and then test print the contents of the specified object.
+   */
+  public static void main(String[] args) throws IOException {
+    // TODO(developer): Replace these variables before running the sample.
+    // The Cloud Storage bucket name.
+    String bucketName = "your-gcs-bucket-name";
+    // The Cloud Storage object name that resides in the specified bucket.
+    String objectName = "your-gcs-object-name";
+
+    System.out.println("Testing token consumer read access...");
+    String content = tokenConsumer(bucketName, objectName);
+    System.out.println(content);
+    System.out.println("done");
+  }
+
+  /** Simulates token broker generating downscoped tokens for specified bucket. */
+  // [START auth_downscoping_token_broker]
+  public static AccessToken getTokenFromBroker(String bucketName, String objectPrefix)
+      throws IOException {
+    // Retrieve the source credentials from ADC.
+    GoogleCredentials sourceCredentials =
+        GoogleCredentials.getApplicationDefault()
+            .createScoped("https://www.googleapis.com/auth/cloud-platform");
+
+    // [START auth_downscoping_rules]
+    // Initialize the Credential Access Boundary rules.
+    String availableResource =
+        String.format("//storage.googleapis.com/projects/_/buckets/%s", bucketName);
+
+    // Downscoped credentials will have readonly access to the resource.
+    String availablePermission = "inRole:roles/storage.objectViewer";
+
+    // Only objects starting with the specified prefix string in the object name will be allowed
+    // read access.
+    String expression =
+        String.format(
+            "resource.name.startsWith('projects/_/buckets/%s/objects/%s')",
+            bucketName, objectPrefix);
+
+    // Define the single access boundary rule using the above properties.
+    CredentialAccessBoundary.AccessBoundaryRule rule =
+        CredentialAccessBoundary.AccessBoundaryRule.newBuilder()
+            .setAvailableResource(availableResource)
+            .addAvailablePermission(availablePermission)
+            .setAvailabilityCondition(
+                CredentialAccessBoundary.AccessBoundaryRule.AvailabilityCondition.newBuilder()
```

Comment:
nit: for clarity it might be nice to have the nested builder created before the AccsesBoundaryRule